From bea2aed0fffc76302d84c879aefe11693d2b7f89 Mon Sep 17 00:00:00 2001
From: Hari Chandrakanthan <quic_haric@quicinc.com>
Date: Tue, 26 Nov 2024 14:27:44 +0530
Subject: [PATCH] QSDK:DP:wifi: ath12k: add support to include fast rx in
 ath12k_dp_rx_process() for ppe vp

Add ppe vp support for utilizing the fast rx support
in ath12k_dp_rx_process().

Also rearrange fields to avoid cache miss
and add a hal ops under peer structure to avoid multiple deference.

This change provides 3-5% cpu improvement.

Remove ATH12K_PEER_STATS_DSBL_SFE and ATH12K_PEER_STATS_DSBL_PPE
as it is not needed.

Patch-depdendency: 971-04-QSDK-ath12k-add-simple-fast-rx-support.patch
Patch-work: None

Signed-off-by: Hari Chandrakanthan <quic_haric@quicinc.com>
---
 drivers/net/wireless/ath/ath12k/debugfs.c | 15 --------------
 drivers/net/wireless/ath/ath12k/dp_rx.c   | 24 +++++++++++++++++++----
 drivers/net/wireless/ath/ath12k/mac.c     | 11 ++---------
 drivers/net/wireless/ath/ath12k/peer.c    | 12 +++---------
 drivers/net/wireless/ath/ath12k/peer.h    | 12 +++++-------
 5 files changed, 30 insertions(+), 44 deletions(-)

diff --git a/drivers/net/wireless/ath/ath12k/debugfs.c b/drivers/net/wireless/ath/ath12k/debugfs.c
index 07d3602..852f1ce 100644
--- a/drivers/net/wireless/ath/ath12k/debugfs.c
+++ b/drivers/net/wireless/ath/ath12k/debugfs.c
@@ -2624,7 +2624,6 @@ static ssize_t ath12k_write_stats_disable(struct file *file,
 	struct ath12k_pdev *pdev;
 	struct ath12k *ar;
 	struct ath12k_link_vif *arvif;
-	struct ath12k_peer *peer;
 	bool disable;
 	int ret, i;
 	u32 mask = 0;
@@ -2636,20 +2635,6 @@ static ssize_t ath12k_write_stats_disable(struct file *file,
 	if (disable != ab->stats_disable) {
 		ab->stats_disable = disable;
 
-		spin_lock_bh(&ab->base_lock);
-		list_for_each_entry(peer, &ab->peers, list) {
-			if ((peer->vif->type == NL80211_IFTYPE_AP) ||
-			    (peer->vif->type == NL80211_IFTYPE_STATION)) {
-				if (peer->ppe_vp_num > 0)
-					peer->stats_disable =
-						ATH12K_PEER_STATS_DSBL_PPE;
-				else
-					peer->stats_disable =
-						ATH12K_PEER_STATS_DSBL_SFE;
-			}
-		}
-		spin_unlock_bh(&ab->base_lock);
-
 		for (i = 0; i < ab->num_radios; i++) {
 			pdev = &ab->pdevs[i];
 			if (pdev && pdev->ar) {
diff --git a/drivers/net/wireless/ath/ath12k/dp_rx.c b/drivers/net/wireless/ath/ath12k/dp_rx.c
index ac578a6..9ea2eff 100644
--- a/drivers/net/wireless/ath/ath12k/dp_rx.c
+++ b/drivers/net/wireless/ath/ath12k/dp_rx.c
@@ -3040,8 +3040,7 @@ static bool ath12k_dp_rx_check_fast_rx(struct ath12k *ar,
 
 	rxcb = ATH12K_SKB_RXCB(msdu);
 
-	if (unlikely(peer->stats_disable != ATH12K_PEER_STATS_DSBL_PPE ||
-	    decap != DP_RX_DECAP_TYPE_ETHERNET2_DIX))
+	if (unlikely(decap != DP_RX_DECAP_TYPE_ETHERNET2_DIX))
 		return false;
 
 	/* mcbc packets go through mac80211 for PN validation */
@@ -3937,8 +3936,7 @@ try_again:
 			rcu_read_lock();
 			peer = ath12k_peer_find_by_peer_list_index(rxcb->ar,
 								   rxcb->peer_id);
-			if (likely(peer &&
-				  (peer->stats_disable == ATH12K_PEER_STATS_DSBL_SFE))) {
+			if (likely(peer && peer->stats_disable)) {
 				intra_bss = u32_get_bits(rxcb->msdu_info,
 							 RX_MSDU_DESC_INFO0_INTRA_BSS);
 				is_mcbc = (u32_get_bits(rxcb->msdu_info,
@@ -3971,6 +3969,24 @@ try_again:
 					msdu->ip_summed =
 						(ip_csum_fail || l4_csum_fail) ?
 						 CHECKSUM_NONE : CHECKSUM_UNNECESSARY;
+
+#if defined(CPTCFG_MAC80211_PPE_SUPPORT) || defined(CPTCFG_ATH12K_PPE_DS_SUPPORT)
+					if (unlikely(peer->ppe_vp_num > 0)) {
+						prefetch(next_msdu->data);
+						rx_desc = (struct hal_rx_desc *)(msdu->data - (hal_rx_desc_sz + l3_pad));
+						peer->hal_ops->rx_desc_get_fse_info(rx_desc, &rx_info);
+						if (!rx_info.flow_idx_timeout &&
+						    !rx_info.flow_idx_invalid &&
+						    (rx_info.flow_metadata != ATH12K_RX_FSE_FLOW_MATCH_SFE)) {
+							msdu->dev = peer->dev;
+							if (ppe_vp_tx_to_ppe(peer->ppe_vp_num, msdu)) {
+								rcu_read_unlock();
+								continue;
+							}
+						}
+					}
+#endif
+
 					msdu->protocol = eth_type_trans(msdu, peer->dev);
 					rcu_read_unlock();
 					prefetch(skb_shinfo(msdu));
diff --git a/drivers/net/wireless/ath/ath12k/mac.c b/drivers/net/wireless/ath/ath12k/mac.c
index 5315a17..eb84f1d 100644
--- a/drivers/net/wireless/ath/ath12k/mac.c
+++ b/drivers/net/wireless/ath/ath12k/mac.c
@@ -7969,16 +7969,9 @@ static void ath12k_sta_set_4addr_wk(struct work_struct *wk)
 			if (peer->vif->type == NL80211_IFTYPE_AP)
 				peer->dev = peer->sta->dev;
 
-			if (ahvif->ppe_vp_type == PPE_VP_USER_TYPE_ACTIVE)
+			if (ahvif->ppe_vp_type == PPE_VP_USER_TYPE_ACTIVE) {
 				peer->ppe_vp_num = ahsta->ppe_vp_num;
-
-			if ((ar->ab->stats_disable) &&
-			    ((peer->vif->type == NL80211_IFTYPE_AP) ||
-			     (peer->vif->type == NL80211_IFTYPE_STATION))) {
-				if (peer->ppe_vp_num > 0)
-					peer->stats_disable = ATH12K_PEER_STATS_DSBL_PPE;
-				else
-					peer->stats_disable = ATH12K_PEER_STATS_DSBL_SFE;
+				peer->hal_ops = ar->ab->hw_params->hal_ops;
 			}
 
 			spin_unlock_bh(&ar->ab->base_lock);
diff --git a/drivers/net/wireless/ath/ath12k/peer.c b/drivers/net/wireless/ath/ath12k/peer.c
index 65ece55..d14f2b6 100644
--- a/drivers/net/wireless/ath/ath12k/peer.c
+++ b/drivers/net/wireless/ath/ath12k/peer.c
@@ -875,22 +875,16 @@ int ath12k_peer_create(struct ath12k *ar, struct ath12k_link_vif *arvif,
 	/* Do not deliver frames to PPE in fast rx incase of RFS
 	 * RFS is supported only in SFE Mode
 	 */
-	if (arvif->ahvif->ppe_vp_type == PPE_VP_USER_TYPE_ACTIVE)
+	if (arvif->ahvif->ppe_vp_type == PPE_VP_USER_TYPE_ACTIVE) {
 		peer->ppe_vp_num = arvif->ahvif->ppe_vp_num;
+		peer->hal_ops = ar->ab->hw_params->hal_ops;
+	}
 #endif
 #ifdef CPTCFG_ATH12K_PPE_DS_SUPPORT
 	if (arvif->ahvif->ppe_vp_type == PPE_VP_USER_TYPE_DS)
 		peer->ppe_vp_num = arvif->ahvif->ppe_vp_num;
 #endif
 
-	if ((ab->stats_disable) && ((peer->vif->type == NL80211_IFTYPE_AP) ||
-	     (peer->vif->type == NL80211_IFTYPE_STATION))) {
-		if (peer->ppe_vp_num > 0)
-			peer->stats_disable = ATH12K_PEER_STATS_DSBL_PPE;
-		else
-			peer->stats_disable = ATH12K_PEER_STATS_DSBL_SFE;
-	}
-
 	if (sta) {
 		ahsta = ath12k_sta_to_ahsta(sta);
 		arsta = ahsta->link[link_id];
diff --git a/drivers/net/wireless/ath/ath12k/peer.h b/drivers/net/wireless/ath/ath12k/peer.h
index 025de53..f32fc60 100644
--- a/drivers/net/wireless/ath/ath12k/peer.h
+++ b/drivers/net/wireless/ath/ath12k/peer.h
@@ -20,15 +20,17 @@ struct ppdu_user_delayba {
 	u32 resp_rate_flags;
 };
 
-#define ATH12K_PEER_STATS_DSBL_SFE	0x1
-#define ATH12K_PEER_STATS_DSBL_PPE	0x2
-
 struct ath12k_peer {
 	struct list_head list;
 	u8 stats_disable;
 	struct net_device *dev;
 	struct ieee80211_sta *sta;
 	struct ieee80211_vif *vif;
+#if defined(CPTCFG_MAC80211_PPE_SUPPORT) || defined(CPTCFG_ATH12K_PPE_DS_SUPPORT)
+	/* Duplicate PPE port number to avoid link vif lookup in rx data path */
+	int ppe_vp_num;
+	const struct hal_ops *hal_ops;
+#endif
 #ifdef CPTCFG_ATH12K_SAWF
 	struct ath12k_sawf_peer_ctx sawf_ctx_peer;
 	struct ath12k_sawf_stats sawf_stats;
@@ -86,10 +88,6 @@ struct ath12k_peer {
 	 */
 	u8 ml_addr[ETH_ALEN];
 	struct rcu_head rcu;
-#if defined(CPTCFG_MAC80211_PPE_SUPPORT) || defined(CPTCFG_ATH12K_PPE_DS_SUPPORT)
-	/* Duplicate PPE port number to avoid link vif lookup in rx data path */
-	int ppe_vp_num;
-#endif
 	bool assoc_success; /* information on peer assoc status from firmware */
 };
 
-- 
2.34.1

